### PR TITLE
feat(goal_planner): change pull over path candidate priority with soft and hard margins

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

change path priority.

before

```
shift (soft) → shift (hard) → arc_forward(soft) → arc_forward (hard) → arc_backward (soft) → arc_backward (hard)
```

after
```
shift(soft1) → shift(soft2) →…→ arc_forward(soft1) → arc_forward (soft1) → … → arc_backward (soft1) → arc_backward (soft2) → … → shift (hard) → arc_forward(hard) → arc_backward (hard)
```

image

image

tier4 internal ticket
https://tier4.atlassian.net/browse/RT1-4623

## Tests performed

psim
https://evaluation.tier4.jp/evaluation/reports/e55d8582-2192-59b9-bbe4-81d8e04d3c81?project_id=prd_jt

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kosuke55** [comment]
this PR include typo fix not related this PR

File: planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 1]
```diff
@@ -706,13 +736,29 @@ std::vector<PullOverPath> GoalPlannerModule::sortPullOverPathCandidatesByGoalPri
...
+  };
+
+  // if object recognition is enabled, sort by collision check margin
   if (parameters_->use_object_recognition) {
+    const std::vector<double> margins = std::invoke([&]() {
+      std::vector<double> combined_margins = soft_margins;
+      combined_margins.insert(combined_margins.end(), hard_margins.begin(), hard_margins.end());
+      return combined_margins;
+    });
+
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp:761]:
```suggestion
    std::vector<double> margins = soft_margins;
    margins.insert(margins.end(), hard_margins.begin(), hard_margins.end());

```
[super nits] simpler?

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!

File: planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp:761]: I like `const`...

**Review by soblin — APPROVED** [review]

**Pull request merged**